vscode-dotty/src/features.ts

```diff
@@ -0,0 +1,64 @@
+import * as vscode from 'vscode'
+import {
+  BaseLanguageClient, ClientCapabilities, DynamicFeature, ServerCapabilities,
+  TextDocumentFeature, TextDocumentRegistrationOptions,
+} from 'vscode-languageclient'
+import { generateUuid } from 'vscode-languageclient/lib/utils/uuid'
+import { DocumentSelector } from 'vscode-languageserver-protocol'
+import { Disposable } from 'vscode-jsonrpc'
+
+import { WorksheetRunRequest } from './protocol'
+import { WorksheetProvider } from './worksheet'
+
+// Remove this if
+// https://github.com/Microsoft/vscode-languageserver-node/issues/423 is fixed.
+function ensure<T, K extends keyof T>(target: T, key: K): T[K] {
+	if (target[key] === void 0) {
+		target[key] = {} as any;
+	}
+	return target[key];
+}
+
+export interface WorksheetClientCapabilities {
+	worksheet?: {
+		run?: {
+			dynamicRegistration?: boolean
+    }
+	}
+}
+
+export interface WorksheetServerCapabilities {
+	/**
+	 * The server provides support for running worksheets.
+	 */
+	worksheetRunProvider?: boolean
+}
+
+export class WorksheetRunFeature extends TextDocumentFeature<TextDocumentRegistrationOptions> {
+	constructor(client: BaseLanguageClient) {
+		super(client, WorksheetRunRequest.type)
+	}
+
+  public fillClientCapabilities(capabilities: ClientCapabilities & WorksheetClientCapabilities): void {
+    ensure(ensure(capabilities, "worksheet")!, "run")!.dynamicRegistration = true
+  }
```

Comment:
`ensure` adds the specified keys to the given object is they're missing, is that it?

Reply:
Yes, I think so, I adapted this from what's done in other features `fillClientCapabilities` in vscode-languageclient